ast/term.go

```diff
@@ -1959,6 +1970,51 @@ func (obj *object) insert(k, v *Term) {
 	obj.ground = obj.ground && k.IsGround() && v.IsGround()
 }
 
+func filterObject(o Value, filter Value) (Value, error) {
+	if filter.Compare(Null{}) == 0 {
+		return o, nil
+	}
+
+	filteredObj, ok := filter.(Object)
+	if !ok {
+		return nil, fmt.Errorf("invalid filter value %q, expected an object", filter)
+	}
+
+	switch v := o.(type) {
+	case String, Number, Boolean, Null:
+		return o, nil
+	case Array:
+		var values Array
+		for i, t := range v {
+			subFilter := filteredObj.Get(StringTerm(strconv.Itoa(i)))
+			if subFilter != nil {
+				filteredValue, err := filterObject(t.Value, subFilter.Value)
+				if err != nil {
+					return nil, err
+				}
+				values = append(values, NewTerm(filteredValue))
+			}
+		}
+		return values, nil
+	case Object:
+		values := NewObject()
+		err := v.Iter(func(k *Term, v *Term) error {
```

Comment:
Nit: One simple optimization would be to make this iterate over the smaller of the two objects.

Reply:
Done.